File: 1927-sum-game/solution.cpp
#include <string>

class Solution {
public:
    bool sumGame(std::string num) {
        int n = num.length();
        int leftSum = 0, rightSum = 0;
        int leftQ = 0, rightQ = 0;

        for (int i = 0; i < n / 2; ++i) {
            if (num[i] == '?') {
                leftQ++;
            } else {
                leftSum += num[i] - '0';
            }
        }

        for (int i = n / 2; i < n; ++i) {
            if (num[i] == '?') {
                rightQ++;
            } else {
                rightSum += num[i] - '0';
            }
        }

        // Bob wins if and only if 2 * (leftSum - rightSum) + 9 * (leftQ - rightQ) == 0.
        // If it equals 0, Bob wins (return false). Otherwise, Alice wins (return true).
        return 2 * (leftSum - rightSum) + 9 * (leftQ - rightQ) != 0;
    }
};

File: 1927-sum-game/solution.py
class Solution:
    def sumGame(self, num: str) -> bool:
        n = len(num)
        half = n // 2
        
        left_sum = 0
        left_q = 0
        for char in num[:half]:
            if char == '?':
                left_q += 1
            else:
                left_sum += int(char)
                
        right_sum = 0
        right_q = 0
        for char in num[half:]:
            if char == '?':
                right_q += 1
            else:
                right_sum += int(char)
                
        # Bob wins if and only if 2 * (left_sum - right_sum) + 9 * (left_q - right_q) == 0.
        # Return True if Alice wins, False if Bob wins.
        return 2 * (left_sum - right_sum) + 9 * (left_q - right_q) != 0